File: apps/desktop/Sources/DahliaRuntimeSupport/TypeID.swift
import Foundation

/// Public representation only; persistence and domain models continue to use UUID.
public enum TypeID {
    public enum Kind: String, CaseIterable, Sendable {
        case vault = "vlt", project = "proj", meeting = "mtg", file, attachment = "att"
        case summary = "sum", transcript, segment = "seg", recording = "rec", event = "evt"
        case summaryJob = "sjob"
        case user, organization = "org", team, organizationMember = "omem", teamMember = "tmem"
        case invitation = "inv", session = "sess", transaction = "txn", operation = "op", patch
    }

    public enum Failure: Error { case invalidID }

    private static let alphabet = Array("0123456789abcdefghjkmnpqrstvwxyz".utf8)

    public static func encode(_ uuid: UUID, as kind: Kind) -> String {
        var bytes = uuid.uuid
        let source = withUnsafeBytes(of: &bytes) { Array($0) }
        var output: [UInt8] = []
        var accumulator = 0
        var bits = 2
        for byte in source {
            accumulator = (accumulator << 8) | Int(byte)
            bits += 8
            while bits >= 5 {
                bits -= 5
                output.append(alphabet[(accumulator >> bits) & 31])
            }
            accumulator &= (1 << bits) - 1
        }
        return kind.rawValue + "_" + String(decoding: output, as: UTF8.self)
    }

    public static func decode(_ value: String, as kind: Kind) throws -> UUID {
        let prefix = kind.rawValue + "_"
        guard value.hasPrefix(prefix) else { throw Failure.invalidID }
        let suffix = Array(value.dropFirst(prefix.count).utf8)
        guard suffix.count == 26, let first = suffix.first, first >= 48, first <= 55 else { throw Failure.invalidID }
        var output: [UInt8] = []
        var accumulator = 0
        var bits = -2
        for character in suffix {
            guard let digit = alphabet.firstIndex(of: character) else { throw Failure.invalidID }
            accumulator = (accumulator << 5) | digit
            bits += 5
            if bits >= 8 {
                bits -= 8
                output.append(UInt8((accumulator >> bits) & 255))
            }
            if bits >= 0 { accumulator &= (1 << bits) - 1 }
        }
        return UUID(uuid: (
            output[0],
            output[1],
            output[2],
            output[3],
            output[4],
            output[5],
            output[6],
            output[7],
            output[8],
            output[9],
            output[10],
            output[11],
            output[12],
            output[13],
            output[14],
            output[15]
        ))
    }
}

File: apps/server/src/typeid.ts
export const idPrefixes = {
  workspace: "ws", project: "proj", meeting: "mtg", file: "file", attachment: "att",
  summary: "sum", transcript: "transcript", segment: "seg", recording: "rec",
  event: "evt", summaryJob: "sjob", contact: "contact", topic: "topic", insight: "inst",
  projectReference: "prr", user: "user", organization: "org", team: "team",
  organizationMember: "omem", teamMember: "tmem", invitation: "inv", session: "sess",
  transaction: "txn", operation: "op", patch: "patch",
} as const;

export type IDKind = keyof typeof idPrefixes;
const alphabet = "0123456789abcdefghjkmnpqrstvwxyz";
const uuidPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function encodeId(kind: IDKind, uuid: string): string {
  if (!uuidPattern.test(uuid)) throw new Error("invalid_uuid");
  let value = BigInt(`0x${uuid.replaceAll("-", "")}`);
  let suffix = "";
  for (let index = 0; index < 26; index += 1) {
    suffix = alphabet[Number(value & 31n)] + suffix;
    value >>= 5n;
  }
  return `${idPrefixes[kind]}_${suffix}`;
}

export function decodeId(kind: IDKind, id: string): string {
  const prefix = `${idPrefixes[kind]}_`;
  if (!id.startsWith(prefix)) throw new Error("invalid_typeid");
  const suffix = id.slice(prefix.length);
  if (!/^[0-7][0-9abcdefghjkmnpqrstvwxyz]{25}$/.test(suffix)) throw new Error("invalid_typeid");
  let value = 0n;
  for (const character of suffix) value = (value << 5n) | BigInt(alphabet.indexOf(character));
  const hex = value.toString(16).padStart(32, "0");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}
